### PR TITLE
[RuntimeEnv] Fix concurrent package download deadlock

### LINKED ISSUE #30369: [Runtime Env] Concurrent working_dir creation will cause agent process to deadlock
### What happened + What you expected to happen

When there is to concurrent working dir creation referencing the same working directory, the concurrent async [`download_and_unpack_package`](https://github.com/ray-project/ray/blob/dcad1a8b7cd7c2c160cd75a61951fb4009b6d7d1/python/ray/_private/runtime_env/packaging.py#L600) functions will deadlock each other. 

https://github.com/ray-project/ray/blob/dcad1a8b7cd7c2c160cd75a61951fb4009b6d7d1/python/ray/_private/runtime_env/packaging.py#L631-L647

This happens when the same URI is being asked to downloaded at the same time, which is common because many workers can share the same working dir but different pip or env var dependencies. 

Looking closer to the call, the lock happen when:
- Coroutine A creates the file lock, and enter into `await internal_kv_get`.
- Coroutine B enters and tries to acquire the file lock, it blocks synchronously.
- Coroutine A can no longer return from the `await` or process, because the file lock is blocked in `acquire`. 

### Versions / Dependencies

2.1.0 and master.

### Reproduction script

```python
import tempfile
from pathlib import Path

import ray
from ray._private import gcs_utils
from ray._private.runtime_env.packaging import (
    get_uri_for_directory,
    upload_package_if_needed,
)
from ray._private.utils import get_directory_size_bytes


def upload_dir(tmpdir):
    gcs_aio_client = gcs_utils.GcsAioClient(
        address=ray.worker.global_worker.gcs_client.address
    )
    # Create an arbitrary nonempty directory to upload.
    path = Path(tmpdir)
    dir_to_upload = path / "dir_to_upload"
    dir_to_upload.mkdir(parents=True)
    filepath = dir_to_upload / "file"
    with filepath.open("w") as file:
        file.write("F" * 100)

    uri = get_uri_for_directory(dir_to_upload)
    assert get_directory_size_bytes(dir_to_upload) > 0

    uploaded = upload_package_if_needed(uri, tmpdir, dir_to_upload)
    assert uploaded
    return uri


ray.init()

tmpdir = tempfile.mkdtemp()
uri = upload_dir(tmpdir)
print(uri)

@ray.remote(runtime_env={"working_dir": uri, "env_vars": {"A": "B"}})
def f():
    print("f")

@ray.remote(runtime_env={"working_dir": uri})
def g():
    print("g")

refs = [f.remote(), g.remote()]
ray.get(refs)
```

This script will hang forever.

You can also go into `/tmp/ray/session_latest/runtime_resources/working_dir_files` to check that there is one single file lock. And if you run `py-spy dump AGENT_PID`, you can see it stuck in `acquire`. 

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by simon-mo**
Signed-off-by: simon-mo <simon.mo@hey.com>

## Why are these changes needed?
See #30369

## Related issue number
Fixes #30369

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by architkulkarni — APPROVED** [review]
Looks good, thanks for the quick fix!

File: python/ray/_private/runtime_env/packaging.py [hunk 1]
```diff
@@ -51,6 +52,24 @@ def _mib_string(num_bytes: float) -> str:
...
+    def __init__(self, lock_file: str):
+        self.file = FileLock(lock_file)
+
+    async def __aenter__(self):
+        while True:
+            acquired = self.file.acquire(blocking=False)
+            if acquired:
+                return
+            else:
+                await asyncio.sleep(0.1)
```
Comment (sihanwang41) [inline python/ray/_private/runtime_env/packaging.py:67]: nit: provide a backoff mechanism.

**Review by sihanwang41 — APPROVED** [review]
nice fix!

**Comment by architkulkarni** [comment]
The test introduced in this PR is in `test_runtime_env_working_dir_2.py` which did not run in CI due to `Post-wheel tests` timing out on earlier tests as a result of the recent `import pandas` breakage on master. So we have to merge master again.  

A different set of tests are broken on the latest master but runtime env tests will run.  Then we can merge this PR

**Comment by simon-mo** [comment]
image

**Pull request merged**